### PR TITLE
docs: Adding CDK example

### PR DESCRIPTION
**Pull request by AlexeySoshin**
**Issue #, if available:**
https://github.com/aws-powertools/powertools-lambda-java/issues/1067

## Description of changes:
Adding CDK examples module to the PowerTools Core example

**Checklist**

* [ ] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [x] Update tests
* [x] Update docs
* [ ] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/powertools-examples-core/cdk/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/cdk/README.md:1]: We're missing a `powertools-examples-core/README.md`

File: examples/powertools-examples-core/sam/src/main/java/helloworld/App.java
Comment (scottgerring) [inline examples/powertools-examples-core/sam/src/main/java/helloworld/App.java:1]: Any idea why these are marked as new files? 
Has anything changed - do we need to diff this externally?

File: examples/powertools-examples-core/cdk/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/cdk/README.md:1]: We're missing a powertools-examples-core/README.md to link down to the concrete examples, and provide probably some high level guidance on which one to consider

File: examples/powertools-examples-core/cdk/app/pom.xml
Comment (scottgerring) [inline examples/powertools-examples-core/cdk/app/pom.xml:3]: What version of Java is the CDK forcing? It might be the case that we should only support Java > 1.8 here if the CDK requires that. @jeromevdl ?

File: examples/powertools-examples-core/cdk/README.md
Comment (scottgerring) [reply examples/powertools-examples-core/cdk/README.md:1]: Also the first part of this doesn't mention the CDK at all. Probably something for the title even?

File: examples/powertools-examples-core/cdk/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/cdk/README.md:58]: Does the SAM stuff belong here?

File: examples/powertools-examples-core/cdk/app/pom.xml
Comment (jeromevdl) [reply examples/powertools-examples-core/cdk/app/pom.xml:3]: You can use different JDK versions for the CDK code and the function code...

File: examples/powertools-examples-core/cdk/README.md
Comment (jeromevdl) [reply examples/powertools-examples-core/cdk/README.md:20]: nop

File: examples/powertools-examples-core/cdk/README.md
Comment (jeromevdl) [reply examples/powertools-examples-core/cdk/README.md:58]: not sure we should mention SAM here... even if we can use both together

File: examples/powertools-examples-core/cdk/app/pom.xml
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/app/pom.xml:3]: I'm happy to bump it to at least Java11, as the Java8 code in CDK is a bit clunky

File: examples/powertools-examples-core/cdk/README.md
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/README.md]: I rewrote this file completely now.

File: examples/powertools-examples-core/cdk/app/pom.xml
Comment (scottgerring) [reply examples/powertools-examples-core/cdk/app/pom.xml:3]: > I'm happy to bump it to at least Java11, as the Java8 code in CDK is a bit clunky



> You can use different JDK versions for the CDK code and the function code...

IMHO given CDK introduces a lower constraint on Java we may as well have the same on the example itself. Then for the build the whole CDK sub-tree only runs for Java-11 plus, or what have you. Otherwise it's almost academic - "yes, you can build the deployable component of this example, but you can't deploy it".

What do you think @jeromevdl ?

File: examples/powertools-examples-core/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/README.md:8]:
```suggestion
We provide examples for the following tools:
```

File: examples/powertools-examples-core/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/README.md:9]:
```suggestion
* [AWS SAM](sam/) 
```

File: examples/powertools-examples-core/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/README.md:10]:
```suggestion
* [AWS CDK](cdk/)
```

File: examples/pom.xml
Comment (jeromevdl) [reply examples/pom.xml:33]: @scottgerring wdyt?

File: examples/powertools-examples-core/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/README.md:12]: I'm not fan of the architecture term. They are just different IaC tools.

File: examples/powertools-examples-core/cdk/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/README.md:15]:
```suggestion
It is a [Maven](https://maven.apache.org/)-based project, so you can open this project with any Maven compatible Java IDE to build and run tests.
```

File: examples/powertools-examples-core/cdk/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/README.md:20]: we are talking about SAM in the CDK project... ?

File: examples/powertools-examples-core/cdk/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/README.md:24]: I would extract the `cdk bootstrap` from here and say it's a prerequisite for deploying cdk applications, and must be run once for each account / region  on which you want to deploy. And then `cdk deploy` is THE command to actually deploy.

Comment (jeromevdl) [reply examples/powertools-examples-core/cdk/README.md:24]: and `cdk bootstrap` needs an account and region I think...

`cdk bootstrap aws://ACCOUNT-NUMBER-1/REGION-1` ([doc](https://docs.aws.amazon.com/cdk/v2/guide/bootstrapping.html))

File: examples/powertools-examples-core/cdk/README.md
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/README.md:30]: do we have multiple stacks in this one ? not accurate I think.

File: examples/powertools-examples-core/cdk/app/.gitignore
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/app/.gitignore:1]: no need of a .gitignore file here

File: examples/powertools-examples-core/cdk/app/src/main/java/helloworld/App.java
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/app/src/main/java/helloworld/App.java:40]: I think it does not respect the checkstyle format. Make sure to update your branch with main and apply the checkstyle formatting in your IDE.

See https://github.com/aws-powertools/powertools-lambda-java/blob/docs/contributing/CONTRIBUTING.md#code-formatting

File: examples/powertools-examples-core/cdk/infra/.gitignore
Comment (jeromevdl) [reply examples/powertools-examples-core/cdk/infra/.gitignore:1]: put the cdk stuff in the root one

File: examples/powertools-examples-core/cdk/infra/cdk.json
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/infra/cdk.json:2]: not sure that make sense... exec:java is to execute the app, but we cannot execute the function like this...

File: examples/powertools-examples-core/cdk/infra/cdk.json
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/infra/cdk.json:22]: also no sure we need all that stuff... 
ec2, ecs, sns, apigw, route53, etc.

Keep only what is related to our app (iam maybe and a few ones)

File: examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java:22]: can we avoid the fixed version number?

File: examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java:42]:
```suggestion
     * Adds API URL to the outputs
```

File: examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java:62]: if you use Java 11, you should have java 11 in your pom.xml for the app.
Or use the `-Dmaven.compiler.source=11...` line 21 (`mvn package`)

File: examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java
Comment (jeromevdl) [inline examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java:67]: you mixed log level and sample rate...

File: examples/powertools-examples-core/sam/.gitignore
Comment (jeromevdl) [inline examples/powertools-examples-core/sam/.gitignore:1]: remove gitignore here

**Review by jeromevdl — CHANGES_REQUESTED** [review]
Please address the comments.

File: examples/pom.xml
Comment (scottgerring) [reply examples/pom.xml:33]: If we can structure it that way and have it still be easy to copy and work with, I think that's good and healthy. 

But - If it makes it "weird" or difficult to e.g. you can't just grab SAM and start working, I'd prefer to copy-paste the code. IMHO the primary goal should be **ease of use** - you can grab this and start working immediately, and a secondary goal should be **code deduplication**. 

I would expect this depends on whether or not you can easily say "Hey {SAM | CDK | SST}, the code you need comes from ..\..\over\here" and have it still work normally.

File: examples/powertools-examples-core/cdk/README.md
Comment (scottgerring) [reply examples/powertools-examples-core/cdk/README.md:24]: > and `cdk bootstrap` needs an account and region I think...
> 
> `cdk bootstrap aws://ACCOUNT-NUMBER-1/REGION-1` ([doc](https://docs.aws.amazon.com/cdk/v2/guide/bootstrapping.html))

Pretty sure it'll use default account / region like other CLI tools if you don't specify and its unambiguous (e.g., there _is_ a default).

File: examples/powertools-examples-core/cdk/app/pom.xml
Comment (scottgerring) [reply examples/powertools-examples-core/cdk/app/pom.xml:3]: > What version of Java is the CDK forcing? It might be the case that we should only support Java > 1.8 here if the CDK requires that. @jeromevdl ?

Answer [here](https://docs.aws.amazon.com/cdk/v2/guide/work-with-cdk-java.html): 

> Java AWS CDK applications require Java 8 (v1.8) or later. We recommend [Amazon Corretto](https://aws.amazon.com/corretto/), but you can use any OpenJDK distribution or [Oracle's JDK](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html). You will also need [Apache Maven](https://maven.apache.org/download.cgi) 3.5 or later. You can also use tools such as Gradle, but the application skeletons generated by the AWS CDK Toolkit are Maven projects.


So we actually have no excuse here to not support 1.8.

File: examples/powertools-examples-core/cdk/infra/cdk.json
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/infra/cdk.json:2]: To the best of my understanding, this is the entrypoint for the CdkApp, not for the HelloWorld function.
And it's a required key

File: examples/powertools-examples-core/cdk/README.md
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/README.md:30]: This is from the original `cdk init`, but you're right, we have just 1 stack, so this is not very interesting.

File: examples/pom.xml
Comment (AlexeySoshin) [reply examples/pom.xml:33]: @scottgerring with CDK it's actually easy to point to the code. I think with SAM it's a bit harder, and I'm not sure about SST yet

File: examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java:62]: This is already set to Java 11:
https://github.com/AlexeySoshin/powertools-lambda-java/blob/eeeaa6dc0b3969cd14a6cc518a8f6a31b261ec3f/examples/powertools-examples-core/cdk/infra/pom.xml#L21

File: examples/powertools-examples-core/cdk/app/src/main/java/helloworld/App.java
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/app/src/main/java/helloworld/App.java]: Done.
We would need to do the same for other examples, though as well:
https://github.com/aws-powertools/powertools-lambda-java/blob/0e76f04f1946dc3cabb842fe4303ac122b67b563/examples/powertools-examples-idempotency/src/main/java/helloworld/App.java#L41-L40

File: examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/infra/src/main/java/cdk/CdkStack.java]: Done

File: examples/.gitignore
Comment (scottgerring) [inline examples/.gitignore:3]:
```suggestion
cdk.out
.m2
```

File: examples/.gitignore
Comment (scottgerring) [inline examples/.gitignore:4]:
```suggestion
.m2
```

File: examples/powertools-examples-core/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/README.md:12]:
```suggestion
For each of the tools, the example application is the same, and consists of the following files:
```

File: examples/powertools-examples-core/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/README.md:8]:
```suggestion
We provide examples for the following infrastructure-as-code tools:
```

File: examples/powertools-examples-core/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/README.md:14]: I thought the code was going to end up in a shared common dir, rather than the subdir of one of the CI tools - was this not practical?

File: examples/powertools-examples-core/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/README.md:16]: link broken (but maybe this will change based on comment above anyway)

File: examples/powertools-examples-core/cdk/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/cdk/README.md:5]:
```suggestion
For general information on the deployed example itself, you can refer to the parent [README](../README.md)
```

File: examples/powertools-examples-core/sam/README.md
Comment (scottgerring) [inline examples/powertools-examples-core/sam/README.md:5]:
```suggestion
For general information on the deployed example itself, you can refer to the parent [README](../README.md)
```

**Comment by scottgerring** [comment]
Sonarqube reliability lint in examples - I have [renamed the sonarqube config file](.sonarcloud.properties) on the batch branch so that we exclude examples from the copy-paste analysis. You might like to do the same here to get past the quality gate. The previous config file name was incorrect for the analysis mode we are using and had no effect.

File: examples/powertools-examples-core/README.md
Comment (AlexeySoshin) [reply examples/powertools-examples-core/README.md:14]: That discussion was resolved, it seems, but I agreed with you that duplicating the sample app for now is a better solution, since SAM doesn't work well otherwise

**Comment by scottgerring** [comment]
> Sonarqube reliability lint in examples - I have [renamed the sonarqube config file](.sonarcloud.properties) on the batch branch so that we exclude examples from the copy-paste analysis. You might like to do the same here to get past the quality gate. The previous config file name was incorrect for the analysis mode we are using and had no effect.

The sonarqube config is now in main

File: examples/powertools-examples-core/cdk/app/pom.xml
Comment (AlexeySoshin) [reply examples/powertools-examples-core/cdk/app/pom.xml:3]: I downgraded CDK examples to JDK8

**Comment by scottgerring** [comment]
Not in your PR, but can you change the example CLI interaction in [examples/README.md](https://github.com/AlexeySoshin/powertools-lambda-java/blob/cdk-example/examples/README.md) to use a different example, as core now has the nested projects?

```bash
# Switch to the directory containing an example for the powertools-core module
$ cd powertools-examples-core
```

powertools-examples-idempotency might be a good one!

**Comment by scottgerring** [comment]
Looking great! I've gone through all the markdown and code again and tested both the deploys. One more point in addition to the comment above (README link), and then we're good to go. 

All the java files in `examples/powertools-core` need 1/ the function headers ([e.g.](https://github.com/aws-powertools/powertools-lambda-java/blob/e8cd87817499d5da08a24858fd6c4495ac5eda7b/examples/powertools-examples-idempotency/src/main/java/helloworld/App.java#L1-L14) and [license-header file](https://github.com/aws-powertools/powertools-lambda-java/blob/main/license-header)) and 2/ to be run through checkstyle (see [here](https://github.com/aws-powertools/powertools-lambda-java/blob/main/CONTRIBUTING.md)). 

I believe we started doing this since you've taken your branch.

Once you've got these two we can merge!

**Review by scottgerring — APPROVED** [review]
LGTM! Will merge after the last build

**Pull request merged**
